src/main/java/org/mockito/internal/MockitoCore.java

```diff
@@ -78,16 +80,25 @@ public boolean isTypeMockable(Class<?> typeToMock) {
     }
 
     public <T> T verify(T mock, VerificationMode mode) {
-        if (mock == null) {
-            throw nullPassedToVerify();
-        }
-        if (!isMock(mock)) {
-            throw notAMockPassedToVerify(mock.getClass());
-        }
         MockingProgress mockingProgress = mockingProgress();
         VerificationMode actualMode = mockingProgress.maybeVerifyLazily(mode);
-        mockingProgress.verificationStarted(new MockAwareVerificationMode(mock, actualMode));
-        return mock;
+
+        Set<VerificationListener> listeners = mockingProgress.verificationListeners();
+        try {
+            if (mock == null) {
+                throw nullPassedToVerify();
+            }
+            if (!isMock(mock)) {
+                throw notAMockPassedToVerify(mock.getClass());
+            }
+            mockingProgress.verificationStarted(new MockAwareVerificationMode(mock, actualMode, listeners));
+            return mock;
+        } catch (MockitoException e) {
+            for (VerificationListener listener : listeners) {
+                listener.onVerificationException(mock, actualMode, e);
+            }
```

Comment:
So this particular loop is now gone, so that leaves us with three loops left.
one for finding the verification listeners, one for verification succes and a final one for verification failure.

If we do the api the way szczepiq  suggets that would leave us with 2. 
@bric3 do you still think it's worth it in that case? and how do you suggest we do this?
have the internal listener be some sort of composite listener that dispatches to all the other verification listeners? but then how would you add the listeners to this composite listener?